cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java

```diff
@@ -200,21 +184,53 @@ private void closeDataset(DatasetCacheKey key, Dataset dataset) {
   }
 
   @Override
+  public void dismissDataset(Dataset dataset) {
+    Preconditions.checkNotNull(dataset);
+    // static datasets cannot be dismissed
+    if (staticDatasets.containsValue(dataset)) {
+      LOG.warn("Attempt to dismiss static dataset {} from dataset cache", dataset);
+      return;
+    }
+    if (txContext == null || !(dataset instanceof TransactionAware)) {
+      dismissSafely(dataset);
```

Comment:
So it seems like only `dismissSafely` will invalid the dataset from the cache, but not in `dismissAfterTx`. Do it miss some of the logic?